src/backend/distributed/utils/maintenanced.c

```diff
@@ -257,6 +263,13 @@ CitusMaintenanceDaemonMain(Datum main_arg)
 		CHECK_FOR_INTERRUPTS();
 
 		/*
+		 * We clear the metadata cache before every iteration because otherwise
+		 * it might contain stale OIDs if deadlock detection runs during a DROP
+		 * EXTENSION followed by a CREATE EXTENSION.
+		 */
+		ClearMetadataCache();
```

Comment:
Hm. There's no cache invalidation and no locking preventing the extension from being dropped and recreated inbetween get_relname_relid() and the relation actually being opened. So that chance always exists, given how the code is currently written - independent of the background daemon.  And that window is widened because get_relname_relid() itself doesn't do invalidation.  But even without that, it could always change after that function returned - you're going to need locks to prevent that.